pkg/importer/dataStream_test.go

```diff
@@ -7,416 +7,404 @@ import (
-	if replacement != nil {
-		qemuOperations = replacement
-		defer func() { qemuOperations = orig }()
+	AfterEach(func() {
+		By("[AfterEach] closing test server")
+		ts.Close()
+	})
+
+	table.DescribeTable("Image, with import source should", func(dest, endpt string, qemuOperations image.QEMUOperations, wantErr bool) {
+		By("Configuring end point")
+		if !strings.HasPrefix(endpt, "http") {
+			endpt = ts.URL + "/" + endpt
+		}
+		By(fmt.Sprintf("end point is %s", endpt))
+		defer os.RemoveAll(dest)
+		By("Replacing QEMU Operations")
+		replaceQEMUOperations(qemuOperations, func() {
+			By("Copying image")
+			err := CopyImage(dest, endpt, "", "")
+			if !wantErr {
+				Expect(err).NotTo(HaveOccurred())
+			} else {
+				Expect(err).To(HaveOccurred())
+			}
+		})
+	},
+		table.Entry("successfully copy local image", filepath.Join(os.TempDir(), "cdi-testcopy"), "tinyCore.iso", NewQEMUAllErrors(), false),
+		table.Entry("expect failure trying to copy non-existing local image", filepath.Join(os.TempDir(), "cdi-testcopy"), "tinyCoreBad.iso", NewQEMUAllErrors(), true),
+		table.Entry("successfully copy streaming image", "cirros-qcow2.raw", "cirros-qcow2.img", NewFakeQEMUOperations(errors.New("should not be called"), nil, nil), false),
+		table.Entry("streaming image qemu validation fails", "cirros-qcow2.raw", "cirros-qcow2.img", NewFakeQEMUOperations(nil, nil, errors.New("invalid image")), true),
+		table.Entry("streaming image qemu convert fails", "cirros-qcow2.raw", "cirros-qcow2.img", NewFakeQEMUOperations(nil, errors.New("exit 1"), nil), true),
+	)
+
+	table.DescribeTable("internal copy", func(r io.Reader, out string, qemu bool, qemuOperations image.QEMUOperations, wantErr bool) {
+		defer os.Remove(out)
+		By("Replacing QEMU Operations")
+		replaceQEMUOperations(qemuOperations, func() {
+			err := copy(r, out, qemu)
+			if !wantErr {
+				Expect(err).NotTo(HaveOccurred())
+			} else {
+				Expect(err).To(HaveOccurred())
+			}
+		})
+	},
+		table.Entry("successfully copy reader", stringRdr, "testoutfile", false, NewFakeQEMUOperations(nil, errors.New("Shouldn't get this"), nil), false),
+		table.Entry("successfully copy qcow2 reader", testFileRdrs, "testqcow2file", true, NewFakeQEMUOperations(nil, errors.New("Shouldn't get this"), nil), false),
+		table.Entry("expect error trying to copy invalid format", testFileRdrs, "testinvalidfile", true, NewFakeQEMUOperations(nil, nil, errors.New("invalid format")), true),
+		table.Entry("expect error trying to copy qemu process fails", testFileRdrs, "testinvalidfile", true, NewFakeQEMUOperations(errors.New("exit 1"), nil, nil), true),
+	)
+})
+
+var _ = Describe("close readers", func() {
+	type args struct {
+		readers []reader
 	}
-	f()
-}
 
-func createDataStream(ep, accKey, secKey string) *DataStream {
-	dsurl, _ := ParseEndpoint(ep)
+	rdrs1 := ioutil.NopCloser(strings.NewReader("test data for reader 1"))
+	rdrs2 := ioutil.NopCloser(strings.NewReader("test data for reader 2"))
+	rdrA := reader{4, rdrs1}
```

Comment:
better to not hard-code the RdrType values.

Reply:
Yeah, was like that in the original file, updated to use the constants instead.